src/main/java/com/compilercharisma/chameleonbusinessstudio/service/AuthenticationService.java

```diff
@@ -40,30 +46,78 @@ public AbstractUser getLoggedInUser(){
             throw new UserNotRegisteredException();
         }
         OAuth2User googleUser = getLoggedInGoogleUser();
-        return users.get(googleUser.getAttribute("email"));
+        return googleUserToAbstractUser(googleUser);
     }
 
     public OAuth2User getLoggedInGoogleUser(){
         if(!isUserLoggedIn()){
             throw new NoUserLoggedInException();
         }
         Object p = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
-        if(!(p instanceof OAuth2User)){
-            throw new RuntimeException();
-        }
-        return (OAuth2User)p;
+        return principalToGoogleUser(p);
+    }
+
+    /**
+     * This method may replace some of our other methods, as it sounds like
+     * SecurityContextHolder is unreliable in Webflux: 
+     * https://stackoverflow.com/q/70050719
+     * 
+     * @return a mono containing the logged in user, or an error if no user is
+     *  logged in
+     */
+    public Mono<UserEntity> getLoggedInUserReactive(){
+        return ReactiveSecurityContextHolder
+            .getContext()
+            .map(SecurityContext::getAuthentication)
+            .handle((auth, sink)->{
+                // https://stackoverflow.com/a/53596358
+                if(isAuthenticatedUser(auth)){
+                    sink.next(auth.getPrincipal());
+                } else {
+                    sink.error(new NoUserLoggedInException());
+                }
+            })
+            .map(this::principalToGoogleUser)
+            .map(this::googleUserToAbstractUser)
+            .map(this::abstractUserToUserEntity);            
+    }
+
+    public Mono<Boolean> isReactiveUserLoggedIn(){
+        return ReactiveSecurityContextHolder
+            .getContext()
+            .map(SecurityContext::getAuthentication)
+            .map(this::isAuthenticatedUser);
+    }
+
+    private boolean isAuthenticatedUser(Authentication auth){
+        return auth != null 
+            && auth.isAuthenticated() 
+            && auth.getPrincipal() != null 
+            && !(auth instanceof AnonymousAuthenticationToken);
     }
```

Comment:
There's no need for boolean logic to figure out if the user is already logged in.

Reply:
Oh, duh! Now I see what you're saying: All of our endpoints are either "unauthenticated can access, don't expect authenticated" or "authenticated only", so an authenticated endpoint only has authenticated users